synapse/rest/client/v1/room.py

```diff
@@ -267,6 +268,226 @@ def on_PUT(self, request, room_id, event_type, txn_id):
         )
 
 
+class RoomBulkSendEventRestServlet(TransactionRestServlet):
+    def __init__(self, hs):
+        super().__init__(hs)
+        self.store = hs.get_datastore()
+        self.state_store = hs.get_storage().state
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.room_member_handler = hs.get_room_member_handler()
+        self.auth = hs.get_auth()
+
+    async def _send_marker_event_for_insertion(
+        self, requester, room_id, insertion_event_id, auth_event_ids
+    ):
+        marker_event_dict = {
+            "type": EventTypes.MSC2716_MARKER,
+            "sender": requester.user.to_string(),
+            "content": {
+                EventContentFields.MSC2716_PREV_INSERTION: insertion_event_id,
+            },
+            "room_id": room_id,
+        }
+
+        (event, _,) = await self.event_creation_handler.create_and_send_nonmember_event(
+            requester,
+            marker_event_dict,
+            # TODO: Do we need to use `self.auth.compute_auth_events(...)` to filter the `auth_event_ids`?
+            auth_event_ids=auth_event_ids,
+        )
+
+    def register(self, http_server):
+        # /rooms/$roomid/bulksend
+        PATTERNS = "/rooms/(?P<room_id>[^/]*)/bulksend"
```

Comment:
Should we move this to an unstable endpoint?

`/_matrix/client/unstable/org.matrix.msc2716/rooms/{roomId}/bulksend`

Reply:
Yes, see https://github.com/matrix-org/synapse/pull/9247#discussion_r652860867